app/src/main/java/org/squidfish/tuplenback/models/GameViewModel.kt

```diff
@@ -208,16 +237,36 @@ class GameViewModel : ViewModel() {
             }
         }
 
-        verifyGame()
+        val game = requireNotNull(game) { throw IllegalStateException("Game cannot be null") }
+
         // end round or end game and get stats
         if (_gameState.value.currentRound >= game.settings.totalRounds) {
             _gameState.update {
                 it.copy(gameOver = true).apply {
                     gameEngines.forEach { (game, gameEngine) ->
-                        it.gameStats[game] = gameEngine.getStats()
+                        if (gameStatsModel[game] == null) {
+                            throw IllegalStateException("Game and GameStats module discrepancy")
+                        }
+
+                        val stats = gameEngine.getStats()
+
+                        gameStatsModel[game]?.let { it1 ->
```

Comment:
If you are giving the argument a name, can it be more meaningful than `it1`. I guess it's because of shadowing warning so choose one `it` to name properly, it doesn't necessarily have to be the inner one

Reply:
Also notice how even though you perform `== null` check above, it is not smart cast here. Another reason to rethink the `throw`/`null` situation